### ISSUE TITLE
K8S use helm3 install yuga 

### ISSUE THREAD

**Issue by likenamehaojie** [yugabyte/yugabyte-db#4964]
![image](https://user-images.githubusercontent.com/9814110/86424342-e1973480-bd14-11ea-9a13-d39834d463a8.png)

**Comment by likenamehaojie** [comment]
?????????????????????????????????????????????????????????????????????????????????????????????????????????????

**Comment by likenamehaojie** [comment]
Failed to initialize sys tables async: Failed log replay. Reason: Failed to open new log: Unable to open readable log segment: /mnt/disk0/yb-data/master/wals/table-sys.catalog.uuid/tablet-00000000000000000000000000000000/wal-000000002: Unable to initialize segment: EOF trying to read 12 bytes at offset 0
    @     0x7f26c666de6c  yb::LogFatalHandlerSink::send()
    @     0x7f26c583f346  google::LogMessage::SendToLog()
    @     0x7f26c583c7aa  google::LogMessage::Flush()
    @     0x7f26c583f879  google::LogMessageFatal::~LogMessageFatal()
    @           0x408ddd  yb::master::MasterMain()
    @     0x7f26c1447825  __libc_start_main
    @           0x4082b9  _start
    @              (nil)  (unknown)
*** Check failure stack trace: ***
W0722 09:04:55.581420    34 scoped_leader_shared_lock.cc:114] Long lock of catalog manager: 0.112s
    @     0x7f26d0568490  yb::master::ScopedLeaderSharedLock::Unlock()
    @     0x7f26d05a8bd1  yb::master::ScopedLeaderSharedLock::~ScopedLeaderSharedLock()
    @     0x7f26d05a59e1  yb::master::MasterServiceImpl::TSHeartbeat()
    @     0x7f26c9e74f96  yb::master::MasterServiceIf::Handle()
    @     0x7f26c7eafcf9  yb::rpc::ServicePoolImpl::Handle()
    @     0x7f26c7e53e04  yb::rpc::InboundCall::InboundCallTask::Run()
    @     0x7f26c7ebb948  yb::rpc::(anonymous namespace)::Worker::Execute()
    @     0x7f26c66fc7af  yb::Thread::SuperviseThread()
    @     0x7f26c1dd7694  start_thread
    @     0x7f26c150d41d  __clone
    @              (nil)  (unknown)
    @     0x7f26c666c251  yb::(anonymous namespace)::DumpStackTraceAndExit()
    @     0x7f26c583cc5d  google::LogMessage::Fail()
    @     0x7f26c583edcd  google::LogMessage::SendToLog()
    @     0x7f26c583c7aa  google::LogMessage::Flush()
    @     0x7f26c583f879  google::LogMessageFatal::~LogMessageFatal()
    @           0x408ddd  yb::master::MasterMain()
    @     0x7f26c1447825  __libc_start_main
    @           0x4082b9  _start
    @              (nil)  (unknown)
*** Aborted at 1595408696 (unix time) try "date -d @1595408696" if you are using GNU date ***
PC: @     0x7f26c145b5a6 __GI_abort
*** SIGSEGV (@0x0) received by PID 1 (TID 0x7f26d0d8ab40) from PID 0; stack trace: ***
    @     0x7f26c1ddfba0 (unknown)
    @     0x7f26c145b5a6 __GI_abort
    @     0x7f26c666c2a4  yb::(anonymous namespace)::DumpStackTraceAndExit()
    @     0x7f26c583cc5c  google::LogMessage::Fail()
    @     0x7f26c583edcc  google::LogMessage::SendToLog()
    @     0x7f26c583c7a9  google::LogMessage::Flush()
    @     0x7f26c583f878  google::LogMessageFatal::~LogMessageFatal()
    @           0x408ddc  yb::master::MasterMain()
    @     0x7f26c1447825 __libc_start_main
    @           0x4082b9 _start
    @                0x0 (unknown)

**Comment by iSignal** [comment]
Hello @likenamehaojie ! Can you share more details about the command line you are using to bring up yugabyte and the environment you are running in? Is this a crash loop or an occasional crash under a workload?

**Comment by kriteshranjan** [comment]
AI validation notes (`yb-issue-4964-analysis.md`):

# Analysis: yugabyte/yugabyte-db#4964

Issue: "K8S use helm3 install yuga" (opened 2020-07-03, open, labels `area/platform`, `community/request`, 3 comments,
last GitHub update 2024-04-02 = metadata only). Analysis date: 2026-09-08. Checkout inspected: `46a9f7adfd` (master,
2026-08-28; shallow clone, so per-tag source was read through the GitHub API).

The title only gives the deployment context (YugabyteDB installed on Kubernetes with Helm 3). The report itself carries
two concrete runtime failures, neither a Helm-3 syntax/compatibility problem:

1. Body screenshot (Windows message box): `FATAL: Not found: Table 000030a7000030008000000000004eb not found in Raft group 00000000000000000000000000000000`
   - a YSQL connection-time FATAL; table id = pg OID 0x4eb = 1259 (`pg_class`) of database OID 0x30a7 = 12455 (an
   initdb-created database), missing from the master sys-catalog tablet.
2. Comment of 2020-07-22: yb-master crash loop
   `Failed to initialize sys tables async: Failed log replay. Reason: Failed to open new log: Unable to open readable log segment: /mnt/disk0/yb-data/master/wals/table-sys.catalog.uuid/tablet-00000000000000000000000000000000/wal-000000002: Unable to initialize segment: EOF trying to read 12 bytes at offset 0`
   (stack ends in `yb::master::MasterMain()` -> `LogMessageFatal`).

The maintainer asked for the command line, environment and crash pattern (2020-07-22); the reporter never answered.

## Verdict
NOT-REPRODUCED — the fully specified failure in this issue (the yb-master crash loop on a zero-length WAL segment,
comment 2020-07-22) cannot be produced on any current target: its producer, the rename-before-header window in
`Log::SwitchToAllocatedSegment`, was closed by `e0afb197a9cb` (#17592/#17760, 2023-06-13) and all five required targets
contain that commit and pass the maintainers' own crash-before-header test protocol (P2/P2b -> P3/P3b = OK on every
target), while the identical protocol on a pre-fix build (2.18.1.0-b84) reproduces the reporter's exact FATAL.
Quick verdict: **Fixed** — the delivery explains the report. Residuals documented below: (a) the screenshot's
`not found in Raft group` class is the still-open #4398 and this issue adds no reproducible information for it;
(b) a Helm-2 `--name` flag survives in one docs example (`helm-chart.md:455`).

## Root cause

### Failure 2 (the crash loop) — producer bug, fixed 2023
- Era code (v2.2.0.0, July 2020): `src/yb/consensus/log.cc:1141-1142` renames the placeholder segment to its final
  `wal-N` name and syncs the directory, and only afterwards `:1166 WriteHeaderAndOpen(header)` writes the 12-byte magic +
  header. A crash (OOM kill, pod eviction, node loss) inside that window leaves `wal-N` with no header. Whether the file
  is zero-length or zero-filled depends on preallocation: `PosixWritableFile::PreAllocate` (v2.2.0.0
  `src/yb/util/env_posix.cc:308-318`, today `src/yb/util/file_system_posix.cc:486-494`) silently returns OK when
  `fallocate` is unsupported (`EOPNOTSUPP`/`ENOSYS`), so on filesystems without fallocate (ZFS < 0.8, several
  network/CSI volumes, cf. #4157) or with `--log_preallocate_segments=false` the file stays 0 bytes.
- Reader side (unchanged since Kudu, still current): `src/yb/consensus/log_util.cc:236-243 ReadableLogSegment::Init()`
  -> `:427-429 ReadHeader()` -> `:468-472 ReadHeaderMagicAndHeaderLength()` -> `env_util::ReadFully`
  (`src/yb/util/env_util.cc:181-184`) returns `IOError "EOF trying to read 12 bytes at offset 0"` for a 0-byte file.
  The zero-filled case is tolerated (`log_util.cc:509-523` `IsAllZeros` -> "treated as a blank segment"), and
  `log_util.cc:414-425 ReadFileSize()` only VLOGs the zero-length case, despite the comment at `:511-513` claiming
  zero-length files are treated the same way.
- Propagation: `log_util.cc:182` ("Unable to initialize segment") -> `src/yb/consensus/log_reader.cc:179-181`
  ("Unable to open readable log segment: <path>") -> `src/yb/consensus/log.cc:754-767 Log::Init` ->
  `src/yb/tablet/tablet_bootstrap.cc:1578-1580` ("Failed to open new log"; `skip_wal_rewrite=true` default at `:112`
  skips the bootstrap-time reader at `:556-558`, so the failure surfaces from `OpenLog`) -> `:584` ("Failed log
  replay. Reason") -> `src/yb/master/catalog_manager.cc:1164-1165` ("Failed to initialize sys tables async") ->
  `src/yb/master/master.cc:394-395` -> `master_main.cc:145` `LOG_AND_RETURN_FROM_MAIN_NOT_OK` =
  `src/yb/util/main_util.h:36-46` `LOG(FATAL)`. Every restart hits the same file: a crash loop until the empty file is
  removed by hand.
- Fix: `e0afb197a9cb` "[#17592 #17760] docdb : Rename WAL file after writing the file header on the disk"
  (2023-06-13, D26156, Jira DB-6732/DB-6856). Current `src/yb/consensus/log.cc:2264-2278`: write header, `Sync()`,
  then rename the hidden `.tmp.newsegmentXXXXXX` placeholder (`log.cc:278-279`, `:2352-2365`) to `wal-N`; hidden
  placeholders are ignored by `log_util.cc:1303-1308 IsLogFileName`. Regression test
  `src/yb/integration-tests/tablet_server-itest.cc:44-74 TabletServerITest.TestCrashBeforeWritingWALHeader`
  (uses `--log_preallocate_segments=false` and `TEST_crash_before_wal_header_is_written`, `log.cc:2261-2263`).
  #17592's own description quotes the reporter's exact message chain.

### Failure 1 (the screenshot) — sibling of open #4398, not recoverable from this report
- Error site: `src/yb/tablet/tablet_metadata.cc:936-959 MakeTableNotFound` (`"Table ... not found in Raft group"`),
  reached when a YSQL backend's catalog read targets a table id absent from the sys-catalog tablet superblock.
  Two mechanisms fit the 2020 evidence: (a) connecting while the master's global initdb was still creating the
  database's catalog tables (in 2020 initdb ran for minutes on Kubernetes; `pg_database` row visible before
  `pg_class` of that database exists), or (b) sys-catalog data loss on the same unreliable storage that later produced
  the zero-length WAL. #4398 (open, 2020-05) collects identical reports from Helm/OpenShift/rook-CephFS deployments;
  the reporter provided no version, logs or timeline, so no faithful reproduction exists for this signature and no
  affected/fixed claim is made for it.

### Title context (Helm 3)
- Helm 3 was already the documented path when the issue was filed: docs at `v2.2.0.0`
  `docs/content/latest/deploy/kubernetes/single-zone/oss/helm-chart.md:62` "Helm 3.0 or later", `:121`
  `helm install yb-demo yugabytedb/yugabyte --namespace yb-demo --wait`; yugabyte/charts `80fce0f527` (2020-03-04,
  "installing yugabytedb chart with helm3"), `d6547e4dab` (2020-04-30), `d022dd6e92` (2020-05-13); YBA Helm 3 via
  #4416 (`346ad7aea0`, 2020-05-02). Today: charts `stable/yugabyte/Chart.yaml` `apiVersion: v2`, `appVersion
  2026.1.1.1-b2`; `docs/content/stable/deploy/kubernetes/single-zone/oss/helm-chart.md:45` "Helm 3.4 or later".
- Residual docs nit: `docs/content/stable/deploy/kubernetes/single-zone/oss/helm-chart.md:455` (and the v2.20,
  v2024.2, v2025.1, v2025.2 copies at `:364`/`:449`) still show `helm install yugabyte ... --name yb-demo ...`, a Helm-2
  flag that Helm 3 rejects (`unknown flag: --name`).

## Trigger matrix
Predicted from source before running; observed with the harness in `## Reproduction` (single yb-master, sys catalog
tablet, `--log_preallocate_segments=false --log_async_preallocate_segments=false`, i.e. the reporter's zero-length
case).

| condition | predicted | observed (all 5 targets) | observed (pre-fix 2.18.1.0-b84 control) |
|---|---|---|---|
| P1 fresh start, then stop | master up, `wal-000000001` present | up; `wal-000000001` present | same |
| P2 restart with `--TEST_crash_before_wal_header_is_written=true` (fix's own flag) | fixed: FATAL "Crash before wal header is written", only a hidden 0-byte `.tmp.newsegment*` left, no new `wal-N`; pre-fix: flag unknown | crash hit; `hidden_placeholders=1`; max wal file unchanged | `ERROR: unknown command line flag` (flag added by the fix) |
| P3 plain restart | fixed: OK | `OK` on all 5 | `OK` (nothing was produced by P2) |
| P2b restart with `--TEST_log_fault_after_segment_allocation_min_replicate_index=1 --reuse_unclosed_segment_threshold_bytes=0` (fault point exists on both eras: after rename on pre-fix `log.cc:1815-1824`, before rename on fixed `log.cc:2221-2226`) | fixed: FATAL "Injected fault", second hidden placeholder, no new `wal-N`; pre-fix: zero-length `wal-N` created by the process itself | fault hit; `hidden_placeholders=2`; no new `wal-N` | fault hit; process left `wal-000000003` of **0 bytes** |
| P3b plain restart (the reporter's crash-loop point) | fixed: OK; pre-fix: reporter's FATAL | `OK` on all 5 | `FATAL ... wal-000000003: Unable to initialize segment: EOF trying to read 12 bytes at offset 0` |
| P4 control: hand-inject zero-length `wal-(max+1)`, restart | FATAL EOF on every version (reader path unchanged; a header-less `wal-N` can only come from outside the process after the fix) | `FATAL_EOF` on all 5 (`master_main.cc:143/144`, `env_util.cc:181`; 2.25.2: `:178`) | `FATAL_EOF` (`env_util.cc:155`) |
| P5 remedy: delete the empty file, restart | OK | `OK` on all 5 | `OK` |

All cells matched prediction. Reader behaviour is identical everywhere; only the producer changed, which is exactly
what the fix claims.

## Reproduction
Host: macOS Apple Silicon, podman machine (arm64 VM, 6 vCPU / 8 GiB). No Kubernetes cluster is needed or used: the
crash is a yb-master WAL-bootstrap crash, so one container running `/home/yugabyte/bin/yb-master` reproduces it.
Files: `cell.sh.txt` (runs inside the image), `run_cell.sh.txt` (host driver), `matrix.log.txt` (combined output).

```sh
WORK=/path/to/dir/containing/cell.sh.txt/and/run_cell.sh.txt
for t in 2024.2.11.0-b36-aarch64 2025.1.4.0-b103-aarch64 2025.2.6.0-b111-aarch64 2026.1.1.1-b2-aarch64 2.25.2.0-b359-aarch64; do
  bash "$WORK/run_cell.sh.txt" "$t"                       # x86 hosts: use the -x86_64 tags
done
bash "$WORK/run_cell.sh.txt" 2.18.1.0-b84 --platform linux/amd64   # pre-fix positive control (amd64-only image)
```
`run_cell.sh.txt` runs, per tag: `podman run --rm --pull=never -v "$WORK/cell.sh.txt:/cell.sh:ro" -v "$WORK/out/<tag>:/out" docker.io/yugabytedb/yugabyte:<tag> bash /cell.sh`
and prints `out/<tag>/summary.txt`. `cell.sh.txt` starts `yb-master --fs_data_dirs=/tmp/m --master_addresses=127.0.0.1:7100 --rpc_bind_addresses=127.0.0.1:7100 --webserver_interface=127.0.0.1 --webserver_port=7000 --logtostderr --log_preallocate_segments=false --log_async_preallocate_segments=false`
through phases P1..P5 above (readiness = "Master server successfully started" + "Becoming Leader" + 4 s so the WAL holds
REPLICATE entries; the fault flag needs `GetMinReplicateIndex() >= 1`, `log_reader.cc:356-369`).

Failure signature to grep: `Unable to initialize segment: EOF trying to read 12 bytes at offset 0`.
Observed full line (pre-fix control, P3b):
`F0908 03:01:18.844002   395 master_main.cc:147] IO error (yb/util/env_util.cc:155): Unable to initialize catalog manager: Failed to initialize sys tables async: Failed log replay. Reason: Failed to open new log: Unable to open readable log segment: /tmp/m/yb-data/master/wals/table-sys.catalog.uuid/tablet-00000000000000000000000000000000/wal-000000003: Unable to initialize segment: EOF trying to read 12 bytes at offset 0`
— identical chain to the issue comment. On the five targets the same chain appears only in P4 (hand-injected file).

Controls: P1/P3/P3b/P5 (negative: normal start succeeds), P4 (positive for the grep/FATAL path on every image),
2.18.1.0-b84 P2b/P3b (era positive control: the process itself produces the zero-length `wal-N` and crash-loops).
Total wall time about 30 s per image; containers are `--rm`, none left running.

Failure 1 (screenshot): not reproducible from the report (no version, logs, or timing); same class as open #4398.

## Versions tested
Resolved at: 2026-09-08T02:05Z (controller) — re-verified 2026-09-08T02:2xZ by this worker.
Resolution evidence: Docker Hub v2 `https://hub.docker.com/v2/repositories/yugabytedb/yugabyte/tags?name=<prefix>&page_size=100&ordering=last_updated`
per prefix (`2024.2.` 87 tags, `2025.1.` 36, `2025.2.` 39, `2026.1.` 12, `2.25.` 9; `2.27.`/`2.29.`/`2026.2` = 0) plus
`/tags/latest` (digest `sha256:3926eedf0ff4b99426fbff533128ff16635ed5697a5cf086dde4fafec28e6f75` == `2026.1.1.1-b2`);
GitHub `repos/yugabyte/yugabyte-db/commits/<v-tag>` for each source tag and `commits/master` (HEAD
`34cae400972627c2b967c378838c27427e891058`, 2026-09-07T22:41:15Z). All values agree with the dispatcher's map; no
substitution. Images run: `docker.io/yugabytedb/yugabyte:<tag>-aarch64` (already present locally); `--version`
inside each container printed the revision listed below.

| target | latest resolved version | source revision | observed repro / control | effective status |
|---|---|---|---|---|
| 2024.2 | 2024.2.11.0-b36 (Hub 2026-08-25) | `04544fd94fb4e50d05994784f444b075a4a5dfdb` | P2 crash -> placeholder only, P3 OK; P2b fault -> placeholder only, P3b OK; P4 injected `wal-000000002` -> FATAL_EOF; P5 OK | fixed |
| 2025.1 | 2025.1.4.0-b103 (Hub 2026-04-20) | `690ac4141e57d060fd834312f396643e8c41f260` | same: P3 OK, P3b OK, P4 FATAL_EOF, P5 OK | fixed |
| 2025.2 | 2025.2.6.0-b111 (Hub 2026-09-04) | `ae22ef35ff65de63303467d95ac3e6d2382feba9` | same: P3 OK, P3b OK, P4 FATAL_EOF, P5 OK | fixed |
| 2026.1 | 2026.1.1.1-b2 (== `latest`, Hub 2026-08-18) | `3b010e021f6027744bf9e016b17e108a2eb24e44` | same: P3 OK, P3b OK, P4 injected `wal-000000004` -> FATAL_EOF, P5 OK | fixed |
| master | preview 2.25.2.0-b359 (Hub 2025-05-20; no newer preview exists) | `03bbc9f31eae5c3fecabd8f0afa5e99211e6080d`; HEAD `34cae400972627c2b967c378838c27427e891058` inspected | preview runtime: P3 OK, P3b OK, P4 FATAL_EOF, P5 OK; HEAD source: fix present (`log.cc:2264-2278`), reader unchanged (`log_util.cc:239-241`) | fixed (preview runtime + HEAD source) |
| reported (2020-07) | version not stated by reporter; era source read at `v2.2.0.0` (`log.cc:1141-1166`) | n/a | not runnable (no aarch64/amd64 image identified for the unknown version); era behaviour demonstrated instead on 2.18.1.0-b84 (`8327449bfbacfffcb28e3031b1e5b8277009fad5`, amd64 emulated): P2b left 0-byte `wal-000000003`, P3b FATAL_EOF (reporter's signature), P4 FATAL_EOF, P5 OK | affected (pre-fix control, not a support target) |

First-fixed release provenance (tag/commit ancestry — not runtime proof, not a backport request):
- Master fix `e0afb197a9cbd3fe8b3f55dbfca1e4397f2650bc` (2023-06-13). `compare/<fix>...<tag>` = `ahead, behind_by=0`
  for v2024.2.11.0, v2025.1.4.0, v2025.2.6.0, v2026.1.1.1, v2.25.2.0 and v2.20.0.0: every in-support branch
  (2024.2, 2025.1, 2025.2, 2026.1, master) contained the fix from its first release; first stable release with it =
  2.20.0.0.
- Out-of-support backports (cherry-picks, verified by the "Rename should happen after writing the header" text in
  `src/yb/consensus/log.cc` per tag): 2.18 -> first in v2.18.2.1 (absent in v2.18.1.0/v2.18.0.1); 2.16 -> first in
  v2.16.6.0 (absent in v2.16.5.0); 2.14 -> first in v2.14.11.0 (absent in v2.14.10.4).

## Proposed fix direction
No DocDB code change is required for the reported crash; it is delivered. Optional hardening and follow-ups:
1. Reader robustness (`src/yb/consensus/log_util.cc:236-243` / `log_reader.cc:173-189`): treat a zero-length **last**
   segment like the all-zeros preallocated case (return `false` from `Init()` and skip it, or delete it) instead of
   failing bootstrap, so a truncating filesystem cannot crash-loop a master/tserver. Keep the FATAL for a zero-length
   non-last segment (that is real data loss). Also make `ReadFileSize()` match its comment. Low priority now that the
   producer cannot create such files.
2. Docs: replace the Helm-2 `--name yb-demo` example at
   `docs/content/stable/deploy/kubernetes/single-zone/oss/helm-chart.md:455` (and versioned copies) with
   `helm install yb-demo yugabytedb/yugabyte --version ... --namespace yb-demo --set storage.master.storageClass=...,storage.tserver.storageClass=... --wait`.
3. Issue hygiene: close #4964 as fixed by #17592 (`e0afb197a9cb`) with a pointer to #4398 for the screenshot class;
   the reporter never supplied the requested details.

## Test strategy
- Existing coverage of the fixed producer: `TabletServerITest.TestCrashBeforeWritingWALHeader`
  (`src/yb/integration-tests/tablet_server-itest.cc:44-74`), run via
  `./yb_build.sh release --cxx-test integration-tests_tablet_server-itest --gtest_filter TabletServerITest.TestCrashBeforeWritingWALHeader`.
- If the optional reader hardening (1) is implemented: add a `log-test.cc` case that creates `wal-000000001` with
  entries plus a zero-length `wal-000000002`, opens `LogReader`, and asserts the reader skips/removes the empty last
  segment and fails only when a zero-length segment precedes a valid one. Public entry points only (files on disk +
  `LogReader::Open`), no test-only production APIs.
- Container-level check (this artifact's harness): P4/P5 of `cell.sh.txt` become OK/OK on every image once (1) lands.

## Confidentiality check
Artifact, scripts and logs contain only public GitHub issue content, public image tags/revisions, loopback addresses
(`127.0.0.1`) and synthetic container paths (`/tmp/m`). No customer names, cluster UUIDs from customer systems, PII,
credentials, internal hostnames or internal links. The reporter and maintainer are referred to by role only; the
screenshot copy (`issue-4964-screenshot.png`) shows only the public FATAL text. Jira ids DB-6732/DB-6856 come from the
public issue bodies of #17592/#17760.

**Comment by kriteshranjan** [comment]
AI validation notes (`cell.sh.txt`):

#!/bin/bash
# yugabyte/yugabyte-db#4964 reproduction cell (runs INSIDE a yugabytedb/yugabyte image).
# Reproduces the master crash loop from the issue's 2020-07-22 comment:
#   "Failed log replay. Reason: Failed to open new log: Unable to open readable log segment:
#    .../tablet-00000000000000000000000000000000/wal-000000002: Unable to initialize segment:
#    EOF trying to read 12 bytes at offset 0"
# Root cause (fixed by e0afb197a9cb, #17592): Log::SwitchToAllocatedSegment renamed the placeholder
# to wal-N BEFORE writing the segment header; a crash in that window (with preallocation off or
# fallocate unsupported) left a zero-length wal-N that ReadableLogSegment::Init cannot open.
#
# Phases (single yb-master, sys catalog tablet only, no tservers needed):
#   P1 baseline start  -> master must come up, WAL dir must contain wal-000000001
#   P2 restart with --TEST_crash_before_wal_header_is_written=true and
#      --reuse_unclosed_segment_threshold_bytes=0 (forces a new segment right after bootstrap)
#      -> master must die at the injected crash point; record what file the crash left behind
#   P3 plain restart   -> FIXED: bootstrap OK (placeholder .tmp.newsegment* ignored)
#                         AFFECTED (pre-fix): FATAL EOF on zero-length wal-N
#   P4 control: inject a zero-length wal-(max+1) by hand and restart -> FATAL EOF expected on every
#      version (reader side unchanged by design; storage-corruption scenario)
#   P5 remedy: delete the injected file and restart -> OK
set -u
BIN=/home/yugabyte/bin
FS=/tmp/m
OUT=/out
WAL=$FS/yb-data/master/wals/table-sys.catalog.uuid/tablet-00000000000000000000000000000000
BASE="--fs_data_dirs=$FS --master_addresses=127.0.0.1:7100 --rpc_bind_addresses=127.0.0.1:7100 --webserver_interface=127.0.0.1 --webserver_port=7000 --logtostderr --log_preallocate_segments=false --log_async_preallocate_segments=false"
SIG='Unable to initialize segment: EOF trying to read 12 bytes at offset 0'
CRASHMSG='Crash before wal header is written'
UPMSG='Master server successfully started'

mkdir -p $OUT $FS
$BIN/yb-master --version > $OUT/version.txt 2>&1
echo "version: $(cat $OUT/version.txt | head -1)" > $OUT/summary.txt

start() { local label=$1; shift; $BIN/yb-master $BASE "$@" > $OUT/$label.log 2>&1 & echo $! > $OUT/$label.pid; }
alive() { kill -0 "$(cat $OUT/$1.pid)" 2>/dev/null; }
wait_up() { # label timeout -> 0 up (started + sys catalog leader elected + 4s settle so the WAL holds REPLICATE entries), 2 died, 1 timeout
  local label=$1 t=$2 i=0
  while [ $i -lt "$t" ]; do
    if grep -q "$UPMSG" $OUT/$label.log 2>/dev/null && grep -q "Becoming Leader" $OUT/$label.log 2>/dev/null; then sleep 4; return 0; fi
    alive "$label" || return 2
    sleep 1; i=$((i+1))
  done; return 1; }
wait_exit() { local label=$1 t=$2 i=0; while alive "$label" && [ $i -lt "$t" ]; do sleep 1; i=$((i+1)); done; alive "$label" && return 1 || return 0; }
stop() { local label=$1 pid; pid=$(cat $OUT/$label.pid); kill -TERM "$pid" 2>/dev/null; if ! wait_exit "$label" 30; then kill -KILL "$pid" 2>/dev/null; wait_exit "$label" 10; fi; }
lswal() { echo "--- WAL dir after $1 ---" >> $OUT/summary.txt; ls -la --time-style=+%T $WAL >> $OUT/summary.txt 2>&1; }
fatal_line() { grep -m1 -E "^F[0-9]{4} " $OUT/$1.log | cut -c1-400; }

# ---------------- P1 baseline ----------------
start p1
wait_up p1 180; r=$?
echo "P1_baseline_start: rc=$r (0=up) wal1_exists=$([ -f $WAL/wal-000000001 ] && echo yes || echo no)" >> $OUT/summary.txt
lswal P1
stop p1

# ---------------- P2 crash before header (maintainers' repro, #17592 test plan) ----------------
start p2 --TEST_crash_before_wal_header_is_written=true --reuse_unclosed_segment_threshold_bytes=0
wait_exit p2 180; r=$?
crashed=$(grep -q "$CRASHMSG" $OUT/p2.log && echo yes || echo no)
maxseg=$(ls $WAL 2>/dev/null | grep -E '^wal-[0-9]+$' | sort | tail -1)
placeholder=$(ls -A $WAL 2>/dev/null | grep -c '^\.tmp\.newsegment')
echo "P2_crash_at_rollover: exited=$([ $r -eq 0 ] && echo yes || echo no) crash_flag_hit=$crashed max_wal_file=$maxseg max_wal_size=$(stat -c %s $WAL/$maxseg 2>/dev/null) hidden_placeholders=$placeholder" >> $OUT/summary.txt
lswal P2

# ---------------- P3 plain restart: the reporter's crash-loop point ----------------
start p3
wait_up p3 180; r=$?
if grep -q "$SIG" $OUT/p3.log; then p3=FATAL_EOF; else p3=$([ $r -eq 0 ] && echo OK || echo "rc=$r"); fi
echo "P3_restart_after_crash: result=$p3 fatal_line=$(fatal_line p3)" >> $OUT/summary.txt
lswal P3
stop p3

# ---------------- P2b/P3b: same window via the fault point that exists on pre-fix builds too ----------
# TEST_log_fault_after_segment_allocation_min_replicate_index: LOG(FATAL) inside SwitchToAllocatedSegment.
#   pre-fix (e.g. v2.18.1.0 log.cc:1815-1824): fault sits AFTER rename-to-wal-N, BEFORE WriteHeader
#   fixed  (master log.cc:2221-2226):          fault sits BEFORE header write AND rename
start p2b --TEST_log_fault_after_segment_allocation_min_replicate_index=1 --reuse_unclosed_segment_threshold_bytes=0
wait_exit p2b 180; r=$?
crashed=$(grep -q "Injected fault" $OUT/p2b.log && echo yes || echo no)
maxseg=$(ls $WAL 2>/dev/null | grep -E '^wal-[0-9]+$' | sort | tail -1)
placeholder=$(ls -A $WAL 2>/dev/null | grep -c '^\.tmp\.newsegment')
echo "P2b_injected_fault_at_rollover: exited=$([ $r -eq 0 ] && echo yes || echo no) fault_hit=$crashed max_wal_file=$maxseg max_wal_size=$(stat -c %s $WAL/$maxseg 2>/dev/null) hidden_placeholders=$placeholder" >> $OUT/summary.txt
lswal P2b
alive p2b && stop p2b
start p3b
wait_up p3b 180; r=$?
if grep -q "$SIG" $OUT/p3b.log; then p3b=FATAL_EOF; else p3b=$([ $r -eq 0 ] && echo OK || echo "rc=$r"); fi
echo "P3b_restart_after_fault: result=$p3b fatal_line=$(fatal_line p3b)" >> $OUT/summary.txt
lswal P3b
alive p3b && stop p3b

# ---------------- P4 control: zero-length wal-(max+1) injected by hand ----------------
# (pre-fix builds leave a zero-length wal-N behind after P2b; remove it first so P4 tests only the injection)
for f in $WAL/wal-*; do [ -s "$f" ] || { echo "P3b_cleanup: removed zero-length leftover $(basename "$f")" >> $OUT/summary.txt; rm -f "$f"; }; done
maxseg=$(ls $WAL | grep -E '^wal-[0-9]+$' | sort | tail -1)
n=$((10#${maxseg#wal-} + 1))
inj=$WAL/wal-$(printf %09d $n)
: > "$inj"
start p4
wait_up p4 120; r=$?
if grep -q "$SIG" $OUT/p4.log; then p4=FATAL_EOF; else p4=$([ $r -eq 0 ] && echo OK || echo "rc=$r"); fi
echo "P4_injected_zero_length_$(basename $inj): result=$p4 fatal_line=$(fatal_line p4)" >> $OUT/summary.txt
lswal P4
alive p4 && stop p4

# ---------------- P5 remedy: remove injected file ----------------
rm -f "$inj"
start p5
wait_up p5 180; r=$?
if grep -q "$SIG" $OUT/p5.log; then p5=FATAL_EOF; else p5=$([ $r -eq 0 ] && echo OK || echo "rc=$r"); fi
echo "P5_after_removing_injected_file: result=$p5" >> $OUT/summary.txt
stop p5
echo "DONE" >> $OUT/summary.txt

**Comment by kriteshranjan** [comment]
AI validation notes (`run_cell.sh.txt`):

#!/bin/bash
# Host-side driver: run cell.sh.txt inside one yugabytedb/yugabyte image (Apple Silicon: -aarch64 tags).
# usage: bash run_cell.sh.txt <tag> [extra podman args...]   e.g. bash run_cell.sh.txt 2026.1.1.1-b2-aarch64
set -u
TAG=$1; shift
# work dir = directory holding this script (pwd -P resolves /tmp -> /private/tmp, which podman-machine on macOS needs)
WORK=${WORK:-$(cd "$(dirname "$0")" && pwd -P)}
OUT=$WORK/out/$TAG
mkdir -p "$OUT" "$WORK/dockercfg"; chmod 777 "$OUT"
[ -f "$WORK/dockercfg/config.json" ] || echo '{}' > "$WORK/dockercfg/config.json"
[ -f "$WORK/auth.json" ] || echo '{}' > "$WORK/auth.json"
export DOCKER_CONFIG=$WORK/dockercfg REGISTRY_AUTH_FILE=$WORK/auth.json   # dodge stale gcloud credHelper
NAME=yb4964-$(echo "$TAG" | tr '.' '-')
podman rm -f "$NAME" >/dev/null 2>&1
echo "=== $TAG  ($(date -u +%FT%TZ 2>/dev/null)) ==="
podman run --rm --name "$NAME" --pull=never "$@" \
  -v "$WORK/cell.sh.txt:/cell.sh:ro" -v "$OUT:/out" \
  "docker.io/yugabytedb/yugabyte:$TAG" bash /cell.sh
echo "--- summary ($TAG) ---"; cat "$OUT/summary.txt"

**Comment by kriteshranjan** [comment]
AI validation notes (`matrix.log.txt`):

yugabyte/yugabyte-db#4964 reproduction matrix -- generated 2026-09-08T03:05Z on Apple Silicon podman (arm64 VM, 6 vCPU / 8 GiB)
harness: cell.sh.txt (runs inside the image) driven by run_cell.sh.txt; images docker.io/yugabytedb/yugabyte:<tag>
2.18.1.0-b84 = pre-fix positive control (amd64 image run with --platform linux/amd64 emulation); NOT a required target
signature grep: 'Unable to initialize segment: EOF trying to read 12 bytes at offset 0'

################ 2024.2.11.0-b36-aarch64 ################
version: version 2024.2.11.0 build 36 revision 04544fd94fb4e50d05994784f444b075a4a5dfdb build_type RELEASE built at 18 Aug 2026 00:58:07 UTC
P1_baseline_start: rc=0 (0=up) wal1_exists=yes
--- WAL dir after P1 ---
total 448
drwxr-xr-x. 2 root root       70 02:58:21 .
drwxr-xr-x. 3 root root       73 02:58:17 ..
-rw-r--r--. 1 root root 24000000 02:58:22 index.000000000
-rw-------. 1 root root   270336 02:58:22 wal-000000001
P2_crash_at_rollover: exited=yes crash_flag_hit=yes max_wal_file=wal-000000001 max_wal_size=270336 hidden_placeholders=1
--- WAL dir after P2 ---
total 280
drwxr-xr-x. 2 root root       99 02:58:26 .
drwxr-xr-x. 3 root root       73 02:58:17 ..
-rw-------. 1 root root        0 02:58:26 .tmp.newsegmentHelwbb
-rw-r--r--. 1 root root 24000000 02:58:26 index.000000000
-rw-------. 1 root root   270336 02:58:22 wal-000000001
P3_restart_after_crash: result=OK fatal_line=
--- WAL dir after P3 ---
total 264
drwxr-xr-x. 2 root root       99 02:58:27 .
drwxr-xr-x. 3 root root       73 02:58:17 ..
-rw-------. 1 root root        0 02:58:26 .tmp.newsegmentHelwbb
-rw-r--r--. 1 root root 24000000 02:58:27 index.000000000
-rw-------. 1 root root   270336 02:58:30 wal-000000001
P2b_injected_fault_at_rollover: exited=yes fault_hit=yes max_wal_file=wal-000000001 max_wal_size=270336 hidden_placeholders=2
--- WAL dir after P2b ---
total 280
drwxr-xr-x. 2 root root      128 02:58:36 .
drwxr-xr-x. 3 root root       73 02:58:17 ..
-rw-------. 1 root root        0 02:58:26 .tmp.newsegmentHelwbb
-rw-------. 1 root root        0 02:58:36 .tmp.newsegmentS6mGOf
-rw-r--r--. 1 root root 24000000 02:58:36 index.000000000
-rw-------. 1 root root   270336 02:58:30 wal-000000001
P3b_restart_after_fault: result=OK fatal_line=
--- WAL dir after P3b ---
total 264
drwxr-xr-x. 2 root root      128 02:58:37 .
drwxr-xr-x. 3 root root       73 02:58:17 ..
-rw-------. 1 root root        0 02:58:26 .tmp.newsegmentHelwbb
-rw-------. 1 root root        0 02:58:36 .tmp.newsegmentS6mGOf
-rw-r--r--. 1 root root 24000000 02:58:37 index.000000000
-rw-------. 1 root root   270336 02:58:40 wal-000000001
P4_injected_zero_length_wal-000000002: result=FATAL_EOF fatal_line=F0908 02:58:45.868216   317 master_main.cc:144] IO error (yb/util/env_util.cc:181): Unable to initialize catalog manager: Failed to initialize sys tables async: Failed log replay. Reason: Failed to open new log: Unable to open readable log segment: /tmp/m/yb-data/master/wals/table-sys.catalog.uuid/tablet-00000000000000000000000000000000/wal-000000002: Unable to initialize segment: EOF trying to re
--- WAL dir after P4 ---
total 264
drwxr-xr-x. 2 root root    126 02:58:45 .
drwxr-xr-x. 3 root root     73 02:58:17 ..
-rw-------. 1 root root      0 02:58:26 .tmp.newsegmentHelwbb
-rw-------. 1 root root      0 02:58:36 .tmp.newsegmentS6mGOf
-rw-------. 1 root root 270336 02:58:40 wal-000000001
-rw-r--r--. 1 root root      0 02:58:45 wal-000000002
P5_after_removing_injected_file: result=OK
DONE
--- full crash / FATAL lines ---
[p2] F0908 02:58:26.591423   102 log.cc:2055] T 00000000000000000000000000000000 P 2aebab73347c477da54bf753cbbcdb17: Crash before wal header is written
[p2b] F0908 02:58:36.695119   220 fault_injection.cc:76] Injected fault: 1.0
[p4] F0908 02:58:45.868216   317 master_main.cc:144] IO error (yb/util/env_util.cc:181): Unable to initialize catalog manager: Failed to initialize sys tables async: Failed log replay. Reason: Failed to open new log: Unable to open readable log segment: /tmp/m/yb-data/master/wals/table-sys.catalog.uuid/tablet-00000000000000000000000000000000/wal-000000002: Unable to initialize segment: EOF trying to read 12 bytes at offset 0

################ 2025.1.4.0-b103-aarch64 ################
version: version 2025.1.4.0 build 103 revision 690ac4141e57d060fd834312f396643e8c41f260 build_type RELEASE built at 02 Apr 2026 20:43:14 UTC
P1_baseline_start: rc=0 (0=up) wal1_exists=yes
--- WAL dir after P1 ---
total 448
drwxr-xr-x. 2 root root       70 02:58:58 .
drwxr-xr-x. 3 root root       73 02:58:56 ..
-rw-r--r--. 1 root root 24000000 02:59:00 index.000000000
-rw-------. 1 root root   274432 02:59:01 wal-000000001
P2_crash_at_rollover: exited=yes crash_flag_hit=yes max_wal_file=wal-000000001 max_wal_size=274432 hidden_placeholders=1
--- WAL dir after P2 ---
total 284
drwxr-xr-x. 2 root root       99 02:59:04 .
drwxr-xr-x. 3 root root       73 02:58:56 ..
-rw-------. 1 root root        0 02:59:04 .tmp.newsegmentjS6Qf5
-rw-r--r--. 1 root root 24000000 02:59:04 index.000000000
-rw-------. 1 root root   274432 02:59:01 wal-000000001
P3_restart_after_crash: result=OK fatal_line=
--- WAL dir after P3 ---
total 268
drwxr-xr-x. 2 root root       99 02:59:05 .
drwxr-xr-x. 3 root root       73 02:58:56 ..
-rw-------. 1 root root        0 02:59:04 .tmp.newsegmentjS6Qf5
-rw-r--r--. 1 root root 24000000 02:59:05 index.000000000
-rw-------. 1 root root   274432 02:59:08 wal-000000001
P2b_injected_fault_at_rollover: exited=yes fault_hit=yes max_wal_file=wal-000000001 max_wal_size=274432 hidden_placeholders=2
--- WAL dir after P2b ---
total 284
drwxr-xr-x. 2 root root      128 02:59:13 .
drwxr-xr-x. 3 root root       73 02:58:56 ..
-rw-------. 1 root root        0 02:59:13 .tmp.newsegmentVtKZBD
-rw-------. 1 root root        0 02:59:04 .tmp.newsegmentjS6Qf5
-rw-r--r--. 1 root root 24000000 02:59:13 index.000000000
-rw-------. 1 root root   274432 02:59:08 wal-000000001
P3b_restart_after_fault: result=OK fatal_line=
--- WAL dir after P3b ---
total 268
drwxr-xr-x. 2 root root      128 02:59:14 .
drwxr-xr-x. 3 root root       73 02:58:56 ..
-rw-------. 1 root root        0 02:59:13 .tmp.newsegmentVtKZBD
-rw-------. 1 root root        0 02:59:04 .tmp.newsegmentjS6Qf5
-rw-r--r--. 1 root root 24000000 02:59:14 index.000000000
-rw-------. 1 root root   274432 02:59:17 wal-000000001
P4_injected_zero_length_wal-000000002: result=FATAL_EOF fatal_line=F0908 02:59:22.754951   307 master_main.cc:143] IO error (yb/util/env_util.cc:181): Unable to initialize catalog manager: Failed to initialize sys tables async: Failed log replay. Reason: Failed to open new log: Unable to open readable log segment: /tmp/m/yb-data/master/wals/table-sys.catalog.uuid/tablet-00000000000000000000000000000000/wal-000000002: Unable to initialize segment: EOF trying to re
--- WAL dir after P4 ---
total 268
drwxr-xr-x. 2 root root    126 02:59:22 .
drwxr-xr-x. 3 root root     73 02:58:56 ..
-rw-------. 1 root root      0 02:59:13 .tmp.newsegmentVtKZBD
-rw-------. 1 root root      0 02:59:04 .tmp.newsegmentjS6Qf5
-rw-------. 1 root root 274432 02:59:17 wal-000000001
-rw-r--r--. 1 root root      0 02:59:22 wal-000000002
P5_after_removing_injected_file: result=OK
DONE
--- full crash / FATAL lines ---
[p2] F0908 02:59:04.517688    98 log.cc:2049] T 00000000000000000000000000000000 P 2708b6cc98644f5fa15d0dc97a381b1c: Crash before wal header is written
[p2b] F0908 02:59:13.631424   211 fault_injection.cc:76] Injected fault: 1.0
[p4] F0908 02:59:22.754951   307 master_main.cc:143] IO error (yb/util/env_util.cc:181): Unable to initialize catalog manager: Failed to initialize sys tables async: Failed log replay. Reason: Failed to open new log: Unable to open readable log segment: /tmp/m/yb-data/master/wals/table-sys.catalog.uuid/tablet-00000000000000000000000000000000/wal-000000002: Unable to initialize segment: EOF trying to read 12 bytes at offset 0

################ 2025.2.6.0-b111-aarch64 ################
version: version 2025.2.6.0 build 111 revision ae22ef35ff65de63303467d95ac3e6d2382feba9 build_type RELEASE built at 02 Sep 2026 16:42:37 UTC
P1_baseline_start: rc=0 (0=up) wal1_exists=yes
--- WAL dir after P1 ---
total 512
drwxr-xr-x. 2 root root       70 02:59:36 .
drwxr-xr-x. 3 root root       73 02:59:33 ..
-rw-r--r--. 1 root root 24000000 02:59:37 index.000000000
-rw-------. 1 root root   274432 02:59:37 wal-000000001
P2_crash_at_rollover: exited=yes crash_flag_hit=yes max_wal_file=wal-000000001 max_wal_size=274432 hidden_placeholders=1
--- WAL dir after P2 ---
total 284
drwxr-xr-x. 2 root root       99 02:59:41 .
drwxr-xr-x. 3 root root       73 02:59:33 ..
-rw-------. 1 root root        0 02:59:41 .tmp.newsegmentVyb90d
-rw-r--r--. 1 root root 24000000 02:59:41 index.000000000
-rw-------. 1 root root   274432 02:59:37 wal-000000001
P3_restart_after_crash: result=OK fatal_line=
--- WAL dir after P3 ---
total 268
drwxr-xr-x. 2 root root       99 02:59:42 .
drwxr-xr-x. 3 root root       73 02:59:33 ..
-rw-------. 1 root root        0 02:59:41 .tmp.newsegmentVyb90d
-rw-r--r--. 1 root root 24000000 02:59:42 index.000000000
-rw-------. 1 root root   274432 02:59:45 wal-000000001
P2b_injected_fault_at_rollover: exited=yes fault_hit=yes max_wal_file=wal-000000001 max_wal_size=274432 hidden_placeholders=2
--- WAL dir after P2b ---
total 284
drwxr-xr-x. 2 root root      128 02:59:51 .
drwxr-xr-x. 3 root root       73 02:59:33 ..
-rw-------. 1 root root        0 02:59:51 .tmp.newsegmentSp9URh
-rw-------. 1 root root        0 02:59:41 .tmp.newsegmentVyb90d
-rw-r--r--. 1 root root 24000000 02:59:51 index.000000000
-rw-------. 1 root root   274432 02:59:45 wal-000000001
P3b_restart_after_fault: result=OK fatal_line=
--- WAL dir after P3b ---
total 268
drwxr-xr-x. 2 root root      128 02:59:52 .
drwxr-xr-x. 3 root root       73 02:59:33 ..
-rw-------. 1 root root        0 02:59:51 .tmp.newsegmentSp9URh
-rw-------. 1 root root        0 02:59:41 .tmp.newsegmentVyb90d
-rw-r--r--. 1 root root 24000000 02:59:52 index.000000000
-rw-------. 1 root root   274432 02:59:55 wal-000000001
P4_injected_zero_length_wal-000000002: result=FATAL_EOF fatal_line=F0908 03:00:01.691879   314 master_main.cc:143] IO error (yb/util/env_util.cc:181): Unable to initialize catalog manager: Failed to initialize sys tables async: Failed log replay. Reason: Failed to open new log: Unable to open readable log segment: /tmp/m/yb-data/master/wals/table-sys.catalog.uuid/tablet-00000000000000000000000000000000/wal-000000002: Unable to initialize segment: EOF trying to re
--- WAL dir after P4 ---
total 268
drwxr-xr-x. 2 root root    126 03:00:01 .
drwxr-xr-x. 3 root root     73 02:59:33 ..
-rw-------. 1 root root      0 02:59:51 .tmp.newsegmentSp9URh
-rw-------. 1 root root      0 02:59:41 .tmp.newsegmentVyb90d
-rw-------. 1 root root 274432 02:59:55 wal-000000001
-rw-r--r--. 1 root root      0 03:00:01 wal-000000002
P5_after_removing_injected_file: result=OK
DONE
--- full crash / FATAL lines ---
[p2] F0908 02:59:41.414322    97 log.cc:2069] T 00000000000000000000000000000000 P 4429c898bb2d43b7a58105242dfe7797: Crash before wal header is written
[p2b] F0908 02:59:51.549579   214 fault_injection.cc:76] Injected fault: 1.0
[p4] F0908 03:00:01.691879   314 master_main.cc:143] IO error (yb/util/env_util.cc:181): Unable to initialize catalog manager: Failed to initialize sys tables async: Failed log replay. Reason: Failed to open new log: Unable to open readable log segment: /tmp/m/yb-data/master/wals/table-sys.catalog.uuid/tablet-00000000000000000000000000000000/wal-000000002: Unable to initialize segment: EOF trying to read 12 bytes at offset 0

################ 2026.1.1.1-b2-aarch64 ################
version: version 2026.1.1.1 build 2 revision 3b010e021f6027744bf9e016b17e108a2eb24e44 build_type RELEASE built at 14 Aug 2026 04:03:14 UTC
P1_baseline_start: rc=0 (0=up) wal1_exists=yes
--- WAL dir after P1 ---
total 512
drwxr-xr-x. 2 root root       70 03:04:05 .
drwxr-xr-x. 3 root root       73 03:04:02 ..
-rw-r--r--. 1 root root 24000000 03:04:06 index.000000000
-rw-------. 1 root root   274432 03:04:06 wal-000000001
P2_crash_at_rollover: exited=yes crash_flag_hit=yes max_wal_file=wal-000000001 max_wal_size=287887 hidden_placeholders=1
--- WAL dir after P2 ---
total 300
drwxr-xr-x. 2 root root       99 03:04:11 .
drwxr-xr-x. 3 root root       73 03:04:02 ..
-rw-------. 1 root root        0 03:04:11 .tmp.newsegments0VPvK
-rw-r--r--. 1 root root 24000000 03:04:11 index.000000000
-rw-------. 1 root root   287887 03:04:10 wal-000000001
P3_restart_after_crash: result=OK fatal_line=
--- WAL dir after P3 ---
total 288
drwxr-xr-x. 2 root root      120 03:04:12 .
drwxr-xr-x. 3 root root       73 03:04:02 ..
-rw-------. 1 root root        0 03:04:11 .tmp.newsegments0VPvK
-rw-r--r--. 1 root root 24000000 03:04:12 index.000000000
-rw-------. 1 root root   287887 03:04:10 wal-000000001
-rw-------. 1 root root     4096 03:04:14 wal-000000002
P2b_injected_fault_at_rollover: exited=yes fault_hit=yes max_wal_file=wal-000000002 max_wal_size=360 hidden_placeholders=2
--- WAL dir after P2b ---
total 304
drwxr-xr-x. 2 root root      149 03:04:20 .
drwxr-xr-x. 3 root root       73 03:04:02 ..
-rw-------. 1 root root        0 03:04:20 .tmp.newsegment2iF6BU
-rw-------. 1 root root        0 03:04:11 .tmp.newsegments0VPvK
-rw-r--r--. 1 root root 24000000 03:04:20 index.000000000
-rw-------. 1 root root   287887 03:04:10 wal-000000001
-rw-------. 1 root root      360 03:04:19 wal-000000002
P3b_restart_after_fault: result=OK fatal_line=
--- WAL dir after P3b ---
total 292
drwxr-xr-x. 2 root root      170 03:04:21 .
drwxr-xr-x. 3 root root       73 03:04:02 ..
-rw-------. 1 root root        0 03:04:20 .tmp.newsegment2iF6BU
-rw-------. 1 root root        0 03:04:11 .tmp.newsegments0VPvK
-rw-r--r--. 1 root root 24000000 03:04:21 index.000000000
-rw-------. 1 root root   287887 03:04:10 wal-000000001
-rw-------. 1 root root      360 03:04:19 wal-000000002
-rw-------. 1 root root     4096 03:04:24 wal-000000003
P4_injected_zero_length_wal-000000004: result=FATAL_EOF fatal_line=F0908 03:04:29.260512   330 master_main.cc:143] IO error (yb/util/env_util.cc:181): Unable to initialize catalog manager: Failed to initialize sys tables async: Failed log replay. Reason: Failed to open new log: Unable to open readable log segment: /tmp/m/yb-data/master/wals/table-sys.catalog.uuid/tablet-00000000000000000000000000000000/wal-000000004: Unable to initialize segment: EOF trying to re
--- WAL dir after P4 ---
total 292
drwxr-xr-x. 2 root root    168 03:04:29 .
drwxr-xr-x. 3 root root     73 03:04:02 ..
-rw-------. 1 root root      0 03:04:20 .tmp.newsegment2iF6BU
-rw-------. 1 root root      0 03:04:11 .tmp.newsegments0VPvK
-rw-------. 1 root root 287887 03:04:10 wal-000000001
-rw-------. 1 root root    360 03:04:19 wal-000000002
-rw-------. 1 root root    360 03:04:28 wal-000000003
-rw-r--r--. 1 root root      0 03:04:29 wal-000000004
P5_after_removing_injected_file: result=OK
DONE
--- full crash / FATAL lines ---
[p2] F0908 03:04:11.013566   106 log.cc:2087] T 00000000000000000000000000000000 P 59a2eb85b5374176b6e8b62b20721620: Crash before wal header is written
[p2b] F0908 03:04:20.136374   227 fault_injection.cc:76] Injected fault: 1.0
[p4] F0908 03:04:29.260512   330 master_main.cc:143] IO error (yb/util/env_util.cc:181): Unable to initialize catalog manager: Failed to initialize sys tables async: Failed log replay. Reason: Failed to open new log: Unable to open readable log segment: /tmp/m/yb-data/master/wals/table-sys.catalog.uuid/tablet-00000000000000000000000000000000/wal-000000004: Unable to initialize segment: EOF trying to read 12 bytes at offset 0

################ 2.25.2.0-b359-aarch64 ################
version: version 2.25.2.0 build 359 revision 03bbc9f31eae5c3fecabd8f0afa5e99211e6080d build_type RELEASE built at 08 May 2025 19:05:35 UTC
P1_baseline_start: rc=0 (0=up) wal1_exists=yes
--- WAL dir after P1 ---
total 512
drwxr-xr-x. 2 root root       70 03:00:14 .
drwxr-xr-x. 3 root root       73 03:00:11 ..
-rw-r--r--. 1 root root 24000000 03:00:16 index.000000000
-rw-------. 1 root root   274432 03:00:16 wal-000000001
P2_crash_at_rollover: exited=yes crash_flag_hit=yes max_wal_file=wal-000000001 max_wal_size=274432 hidden_placeholders=1
--- WAL dir after P2 ---
total 284
drwxr-xr-x. 2 root root       99 03:00:20 .
drwxr-xr-x. 3 root root       73 03:00:11 ..
-rw-------. 1 root root        0 03:00:20 .tmp.newsegmentRZ1MKJ
-rw-r--r--. 1 root root 24000000 03:00:20 index.000000000
-rw-------. 1 root root   274432 03:00:16 wal-000000001
P3_restart_after_crash: result=OK fatal_line=
--- WAL dir after P3 ---
total 268
drwxr-xr-x. 2 root root       99 03:00:21 .
drwxr-xr-x. 3 root root       73 03:00:11 ..
-rw-------. 1 root root        0 03:00:20 .tmp.newsegmentRZ1MKJ
-rw-r--r--. 1 root root 24000000 03:00:21 index.000000000
-rw-------. 1 root root   274432 03:00:25 wal-000000001
P2b_injected_fault_at_rollover: exited=yes fault_hit=yes max_wal_file=wal-000000001 max_wal_size=274432 hidden_placeholders=2
--- WAL dir after P2b ---
total 284
drwxr-xr-x. 2 root root      128 03:00:30 .
drwxr-xr-x. 3 root root       73 03:00:11 ..
-rw-------. 1 root root        0 03:00:20 .tmp.newsegmentRZ1MKJ
-rw-------. 1 root root        0 03:00:30 .tmp.newsegmentwnp9vK
-rw-r--r--. 1 root root 24000000 03:00:30 index.000000000
-rw-------. 1 root root   274432 03:00:25 wal-000000001
P3b_restart_after_fault: result=OK fatal_line=
--- WAL dir after P3b ---
total 268
drwxr-xr-x. 2 root root      128 03:00:31 .
drwxr-xr-x. 3 root root       73 03:00:11 ..
-rw-------. 1 root root        0 03:00:20 .tmp.newsegmentRZ1MKJ
-rw-------. 1 root root        0 03:00:30 .tmp.newsegmentwnp9vK
-rw-r--r--. 1 root root 24000000 03:00:31 index.000000000
-rw-------. 1 root root   274432 03:00:35 wal-000000001
P4_injected_zero_length_wal-000000002: result=FATAL_EOF fatal_line=F0908 03:00:40.698760   318 master_main.cc:143] IO error (yb/util/env_util.cc:178): Unable to initialize catalog manager: Failed to initialize sys tables async: Failed log replay. Reason: Failed to open new log: Unable to open readable log segment: /tmp/m/yb-data/master/wals/table-sys.catalog.uuid/tablet-00000000000000000000000000000000/wal-000000002: Unable to initialize segment: EOF trying to re
--- WAL dir after P4 ---
total 268
drwxr-xr-x. 2 root root    126 03:00:40 .
drwxr-xr-x. 3 root root     73 03:00:11 ..
-rw-------. 1 root root      0 03:00:20 .tmp.newsegmentRZ1MKJ
-rw-------. 1 root root      0 03:00:30 .tmp.newsegmentwnp9vK
-rw-------. 1 root root 274432 03:00:35 wal-000000001
-rw-r--r--. 1 root root      0 03:00:40 wal-000000002
P5_after_removing_injected_file: result=OK
DONE
--- full crash / FATAL lines ---
[p2] F0908 03:00:20.393479   102 log.cc:2022] T 00000000000000000000000000000000 P 0689e26758de408b9c2736f5ade1cd3f: Crash before wal header is written
[p2b] F0908 03:00:30.554303   219 fault_injection.cc:76] Injected fault: 1.0
[p4] F0908 03:00:40.698760   318 master_main.cc:143] IO error (yb/util/env_util.cc:178): Unable to initialize catalog manager: Failed to initialize sys tables async: Failed log replay. Reason: Failed to open new log: Unable to open readable log segment: /tmp/m/yb-data/master/wals/table-sys.catalog.uuid/tablet-00000000000000000000000000000000/wal-000000002: Unable to initialize segment: EOF trying to read 12 bytes at offset 0

################ 2.18.1.0-b84 ################
version: version 2.18.1.0 build 84 revision 8327449bfbacfffcb28e3031b1e5b8277009fad5 build_type RELEASE built at 16 Jun 2023 17:36:56 UTC
P1_baseline_start: rc=0 (0=up) wal1_exists=yes
--- WAL dir after P1 ---
total 256
drwxr-xr-x. 2 root root       70 03:01:01 .
drwxr-xr-x. 3 root root       73 03:00:57 ..
-rw-r--r--. 1 root root 24000000 03:01:04 index.000000000
-rw-------. 1 root root   262144 03:01:05 wal-000000001
P2_crash_at_rollover: exited=yes crash_flag_hit=no max_wal_file=wal-000000001 max_wal_size=262144 hidden_placeholders=0
--- WAL dir after P2 ---
total 256
drwxr-xr-x. 2 root root       70 03:01:01 .
drwxr-xr-x. 3 root root       73 03:00:57 ..
-rw-r--r--. 1 root root 24000000 03:01:04 index.000000000
-rw-------. 1 root root   262144 03:01:05 wal-000000001
P3_restart_after_crash: result=OK fatal_line=
--- WAL dir after P3 ---
total 260
drwxr-xr-x. 2 root root       91 03:01:07 .
drwxr-xr-x. 3 root root       73 03:00:57 ..
-rw-r--r--. 1 root root 24000000 03:01:07 index.000000000
-rw-------. 1 root root   262144 03:01:05 wal-000000001
-rw-------. 1 root root     4096 03:01:11 wal-000000002
P2b_injected_fault_at_rollover: exited=yes fault_hit=yes max_wal_file=wal-000000003 max_wal_size=0 hidden_placeholders=0
--- WAL dir after P2b ---
total 276
drwxr-xr-x. 2 root root      112 03:01:16 .
drwxr-xr-x. 3 root root       73 03:00:57 ..
-rw-r--r--. 1 root root 24000000 03:01:16 index.000000000
-rw-------. 1 root root   262144 03:01:05 wal-000000001
-rw-------. 1 root root     4096 03:01:11 wal-000000002
-rw-------. 1 root root        0 03:01:16 wal-000000003
P3b_restart_after_fault: result=FATAL_EOF fatal_line=F0908 03:01:18.844002   395 master_main.cc:147] IO error (yb/util/env_util.cc:155): Unable to initialize catalog manager: Failed to initialize sys tables async: Failed log replay. Reason: Failed to open new log: Unable to open readable log segment: /tmp/m/yb-data/master/wals/table-sys.catalog.uuid/tablet-00000000000000000000000000000000/wal-000000003: Unable to initialize segment: EOF trying to re
--- WAL dir after P3b ---
total 260
drwxr-xr-x. 2 root root     89 03:01:18 .
drwxr-xr-x. 3 root root     73 03:00:57 ..
-rw-------. 1 root root 262144 03:01:05 wal-000000001
-rw-------. 1 root root   4096 03:01:11 wal-000000002
-rw-------. 1 root root      0 03:01:16 wal-000000003
P3b_cleanup: removed zero-length leftover wal-000000003
P4_injected_zero_length_wal-000000003: result=FATAL_EOF fatal_line=F0908 03:01:20.258229   480 master_main.cc:147] IO error (yb/util/env_util.cc:155): Unable to initialize catalog manager: Failed to initialize sys tables async: Failed log replay. Reason: Failed to open new log: Unable to open readable log segment: /tmp/m/yb-data/master/wals/table-sys.catalog.uuid/tablet-00000000000000000000000000000000/wal-000000003: Unable to initialize segment: EOF trying to re
--- WAL dir after P4 ---
total 260
drwxr-xr-x. 2 root root     89 03:01:19 .
drwxr-xr-x. 3 root root     73 03:00:57 ..
-rw-------. 1 root root 262144 03:01:05 wal-000000001
-rw-------. 1 root root   4096 03:01:11 wal-000000002
-rw-r--r--. 1 root root      0 03:01:19 wal-000000003
P5_after_removing_injected_file: result=OK
DONE
--- full crash / FATAL lines ---
[p2] ERROR: unknown command line flag 'TEST_crash_before_wal_header_is_written'
[p2b] F0908 03:01:16.768114   339 fault_injection.cc:76] Injected fault: 1.0
[p3b] F0908 03:01:18.844002   395 master_main.cc:147] IO error (yb/util/env_util.cc:155): Unable to initialize catalog manager: Failed to initialize sys tables async: Failed log replay. Reason: Failed to open new log: Unable to open readable log segment: /tmp/m/yb-data/master/wals/table-sys.catalog.uuid/tablet-00000000000000000000000000000000/wal-000000003: Unable to initialize segment: EOF trying to read 12 bytes at offset 0
[p4] F0908 03:01:20.258229   480 master_main.cc:147] IO error (yb/util/env_util.cc:155): Unable to initialize catalog manager: Failed to initialize sys tables async: Failed log replay. Reason: Failed to open new log: Unable to open readable log segment: /tmp/m/yb-data/master/wals/table-sys.catalog.uuid/tablet-00000000000000000000000000000000/wal-000000003: Unable to initialize segment: EOF trying to read 12 bytes at offset 0

**Comment by kriteshranjan** [comment]
Closing this issue based on AI validation. cc: @m-iancu , @VamsidharGunturu
